scipy/stats/_distribution_infrastructure.py

```diff
@@ -3613,69 +3650,103 @@ def _cdf2_quadrature(self, x, y, **params):
         x_, y_ = np.floor(x), np.floor(y)
-        logcdf_ymx = super()._logcdf2_subtraction(x_, y_, **params)
+        logcdf_ymx = func(x_, y_, **params)
         logpmf_x = np.where(x_ == x, self._logpmf_dispatch(x_, **params), -np.inf)
         return special.logsumexp([logcdf_ymx, logpmf_x], axis=0)
 
+    def _cdf2_subtraction(self, x, y, **params):
+        return self._cdf2_subtraction_base(x, y, super()._cdf2_subtraction, **params)
+
+    def _cdf2_subtraction_safe(self, x, y, **params):
+        return self._cdf2_subtraction_base(x, y, super()._cdf2_subtraction_safe,
+                                           **params)
+
+    def _logcdf2_subtraction(self, x, y, **params):
+        return self._logcdf2_subtraction_base(x, y, super()._logcdf2_subtraction,
+                                              **params)
+
+    def _logcdf2_subtraction_safe(self, x, y, **params):
+        return self._logcdf2_subtraction_base(x, y, super()._logcdf2_subtraction,
+                                              **params)
+
     def _ccdf2_addition(self, x, y, **params):
         a, _ = self._support(**params)
         ccdf_y = self._ccdf_dispatch(y, **params)
         _cdf, args = _kwargs2args(self._cdf_dispatch, kwargs=params)
-        cdf_xm1 = apply_where(x - 1 >= a, [x - 1] + args, _cdf, fillvalue=0)
+        cdf_xm1 = apply_where(x - 1 >= a, (x - 1,) + args, _cdf, fill_value=0)
         return ccdf_y + cdf_xm1
 
     def _logccdf2_addition(self, x, y, **params):
         a, _ = self._support(**params)
         logccdf_y = self._logccdf_dispatch(y, **params)
         _logcdf, args = _kwargs2args(self._logcdf_dispatch, kwargs=params)
-        logcdf_xm1 = apply_where(x - 1 >= a, [x - 1] + args, _logcdf, fillvalue=-np.inf)
+        logcdf_xm1 = apply_where(x - 1 >= a, (x - 1,) + args, _logcdf, fill_value=-np.inf)
         return special.logsumexp([logccdf_y, logcdf_xm1], axis=0)
 
+    def _base_discrete_inversion(self, p, func, comp, /, **params):
+        # For discrete distributions, icdf(p) is defined as the minimum n
+        # such that cdf(n) >= p. iccdf(p) is defined as the minimum n such
+        # that ccdf(n) <= p, or equivalently as iccdf(p) = icdf(1 - p).
+
+        res = self._solve_bounded(func, p, params=params, xatol=0.9)
+        # First try to find where cdf(x) == p for the continuous extension of the
+        # cdf. res.xl and res.xr will be a bracket for this root. The parameter
+        # xatol in solve_bounded controls the bracket width. We thus know that
+        # know cdf(res.xr) >= p, cdf(res.xl) <= p, and |res.xr - res.xl| <= 0.9.
+        # This means the minimum integer n such that cdf(n) >= p is either floor(x)
+        # or floor(x) + 1.
+        res, fr = np.asarray(np.floor(res.xr)), res.fr
+        # xr is a bracket endpoint, and will usually be a finite value even when
+        # the computed result should be nan. We need to explicitly handle this
+        # case.
+        res[np.isnan(fr)] = np.nan
```

Comment:
We'll have to take a look at how many other places this shows up. If it's all over, maybe we do take care of this in the decorator after all.